disttask/framework/scheduler/scheduler.go

```diff
@@ -116,67 +118,101 @@ func (s *InternalSchedulerImpl) Run(ctx context.Context, task *proto.Task) error
 			break
 		}
+		s.subtaskWg.Add(1)
+		s.runSubtask(runCtx, scheduler, subtask, task.Step, minimalTaskCh)
+		if !concurrentSubtask {
+			s.subtaskWg.Wait()
+		}
+	}
+	s.subtaskWg.Wait()
+	return s.getError()
+}
 
-		var minimalTasks []proto.MinimalTask
-		minimalTasks, err = scheduler.SplitSubtask(runCtx, subtask.Meta)
-		if err != nil {
-			s.onError(err)
-			if errors.Cause(err) == context.Canceled {
-				s.updateSubtaskStateAndError(subtask.ID, proto.TaskStateCanceled, "")
-			} else {
-				s.updateSubtaskStateAndError(subtask.ID, proto.TaskStateFailed, s.getError().Error())
-			}
-			break
+func (s *InternalSchedulerImpl) runSubtask(ctx context.Context, scheduler Scheduler, subtask *proto.Subtask, step int64, minimalTaskCh chan func()) {
+	minimalTasks, err := scheduler.SplitSubtask(ctx, subtask.Meta)
+	if err != nil {
+		s.onError(err)
+		if errors.Cause(err) == context.Canceled {
+			s.updateSubtaskStateAndError(subtask.ID, proto.TaskStateCanceled, "")
+		} else {
+			s.updateSubtaskStateAndError(subtask.ID, proto.TaskStateFailed, s.getError().Error())
 		}
-		logutil.Logger(s.logCtx).Info("split subTask", zap.Any("cnt", len(minimalTasks)), zap.Any("subtask_id", subtask.ID))
-		for _, minimalTask := range minimalTasks {
-			minimalTaskWg.Add(1)
-			j := minimalTask
-			minimalTaskCh <- func() {
-				s.runMinimalTask(runCtx, j, task.Type, task.Step)
-				minimalTaskWg.Done()
+		s.subtaskWg.Done()
+		return
+	}
+	logutil.Logger(s.logCtx).Info("split subTask", zap.Any("cnt", len(minimalTasks)), zap.Any("subtask_id", subtask.ID))
+
+	// fast path for ADD INDEX.
+	// ADD INDEX is a special case now, no minimal tasks will be generated.
+	// run it synchronously now.
+	if len(minimalTasks) == 0 {
+		s.onSubtaskFinished(ctx, scheduler, subtask)
+		s.subtaskWg.Done()
+		return
+	}
+
+	var mu sync.Mutex
+	var cnt int
+	for _, minimalTask := range minimalTasks {
+		j := minimalTask
+		minimalTaskCh <- func() {
+			s.runMinimalTask(ctx, j, subtask.Type, step)
+
+			mu.Lock()
+			defer mu.Unlock()
+			cnt++
+			// last minimal task should mark subtask as finished
+			if cnt == len(minimalTasks) {
+				s.onSubtaskFinished(ctx, scheduler, subtask)
+				s.subtaskWg.Done()
 			}
 		}
-		minimalTaskWg.Wait()
-		if err := s.getError(); err == nil {
-			if err := scheduler.OnSubtaskFinished(runCtx, subtask.Meta); err != nil {
-				s.onError(err)
-			}
+	}
+}
+
+func (s *InternalSchedulerImpl) onSubtaskFinished(ctx context.Context, scheduler Scheduler, subtask *proto.Subtask) {
+	if err := s.getError(); err == nil {
+		if err := scheduler.OnSubtaskFinished(ctx, subtask.Meta); err != nil {
+			s.onError(err)
 		}
-		if err := s.getError(); err != nil {
-			if errors.Cause(err) == context.Canceled {
-				s.updateSubtaskStateAndError(subtask.ID, proto.TaskStateCanceled, "")
-			} else {
-				s.updateSubtaskStateAndError(subtask.ID, proto.TaskStateFailed, s.getError().Error())
-			}
-			break
+	}
+	if err := s.getError(); err != nil {
```

Comment:
```suggestion
	} else {
```

Reply:
if OnSubtaskFinished return error(L204), we also need to update subtask status to failed